typescript-ignorant/TypeScriptParser.g4

```diff
@@ -0,0 +1,838 @@
+    ;
+
+enumBody
+    : enumMemberList ','?
+    ;
+
+enumMemberList
+    : enumMember (',' enumMember)*
+    ;
+
+enumMember
+    : propertyName ('=' singleExpression)?
+    ;
+
+// A.8 Namespaces
+
+namespaceDeclaration
+    : Declare? Namespace namespaceName '{' statementList '}'
+    ;
+
+namespaceName
+    : Identifier ('.' Identifier)*?
+    ;
+
+// Modules look just like namespaces to me
+
+moduleDeclaration
+    : Declare? Module moduleName '{' statementList '}'
+    ;
+
+moduleName
+    : Identifier ('.' Identifier)*?
+    ;
+
+importAliasDeclaration
+    : Identifier '=' (namespaceName | moduleName) SemiColon
+    ;
+
+// Ext.2 Additions to 1.8: Decorators
+
+decoratorList
+    : decorator+ ;
+
+decorator
+    : '@' (decoratorMemberExpression | decoratorCallExpression)
+    ;
+
+decoratorMemberExpression
+    : Identifier
+    | decoratorMemberExpression '.' identifierName
+    | '(' singleExpression ')'
+    ;
+
+decoratorCallExpression
+    : decoratorMemberExpression arguments;
+
+// ECMAPart
+program
+    : statementList EOF
+    ;
+
+statement
+    : block                        # StatementBlock
+    | variableStatement            # StatementVariable
+    | emptyStatement               # StatementEmpty
+    | abstractDeclaration          # StatementAbstractDeclaration    //ADDED
+    | classDeclaration             # StatementClassDeclaration
+    | interfaceDeclaration         # StatementInterfaceDeclaration   //ADDED
+    | namespaceDeclaration         # StatementNamespaceDeclaration   //ADDED
+    | moduleDeclaration            # StatementModuleDeclaration      //ADDED
+    | typeAliasDeclaration         # StatementTypeAliasDeclaration   //ADDED
+    | enumDeclaration              # StatementEnumDeclaration        //ADDED
+    | ifStatement                  # StatementIf
+    | iterationStatement           # StatementIteration
+    | continueStatement            # StatementContinue
+    | breakStatement               # StatementBreak
+    | returnStatement              # StatementReturn
+    | yieldStatement               # StatementYield
+    | withStatement                # StatementWith
+    | labelledStatement            # StatementLabelled
+    | switchStatement              # StatementSwitch
+    | throwStatement               # StatementThrow
+    | tryStatement                 # StatementTry
+    | debuggerStatement            # StatementDebugger
+    | functionDeclaration          # StatementFunctionDeclaration
+    | arrowFunctionDeclaration     # StatementArrowFunctionDeclaration
+    | generatorFunctionDeclaration # StatementGeneratorFunctionDeclaration
+    | expressionStatement          # StatementExpression
+    | importStatement              # StatementImport
+    | exportStatement              # StatementExport
+//    | Export statement // exportStatement should cover this
+    ;
+
+block
+    : '{' statementList '}'
+    ;
+
+statementList
+    : statement*?
```

Comment:
I think it bad syntax because it can match an empty string (if you removes  `?`, such a warning will occur). Use '+' here instead of '*' and place an optional symbol to superior rules where necessary.

Also, place `statementList` declaration before the rule `statement`.

Reply:
How is an empty string is different from an empty statement or an absence of a statement?

Reply:
There is a difference between the following constructions:

```g4
a : a b;
b : 't'?;
```

```g4
a : a b?;
b : 't';
```

Always use the second syntax everywhere because of performance problems in the first case, more details on [stackoverflow](https://stackoverflow.com/a/26041612/1046374).

Reply:
From the StackOverflow answer:
> rule _rule_ contains an **optional block** with at least one alternative that can match an empty string

(bold emphasis is mine).

In this case in parent rules the `statementList` block is _not_ optional, thus it imposes no problem.